designs/aws/aws-launch-templates-v2.md

```diff
@@ -0,0 +1,146 @@
+
+If additional configuration is required (say to use a specific AMI), we ask that a complete LaunchTemplate is provided to the Karpenter provisioner. This route has had a few advantages -
+
+1. Users can fully customize their worker node and use any available EC2 feature that’s exposed via launch templates.
+2. Partially filled out launch templates are difficult to interpret. Anyone specifying a custom AMI to use will also need to explicitly specify how it needs to be bootstrapped via UserData. This is important because Karpenter cannot make guesses on how to start the kubelet on an unknown AMI.
+
+The main, critical downside of this approach has been that a user needs to maintain a fully-formed launch template even when they’re only looking to customize a single property of their worker nodes. Consider these examples -
+
+1. In order to override [the root volume](https://github.com/aws/karpenter/issues/885), you’d need to author an entire launch template containing security groups, userData etc.
+2. In order to just have all statically provisioned [volumes be encrypted](https://github.com/aws/karpenter/issues/933), you need to author an entire launch template.
+
+While some users do like complete control over the launch template since they customize everything (kOps), most only need to override a few settings and would therefore like those to be present as knobs on the provisioner itself, and for all other settings to be figured out by Karpenter.
+
+
+### **Which parameters should we support in the providerSpec?**
+
+The reason this question is important is because we don’t want to bloat the AWS providerSpec by adding every EC2 property in there. Having said that, we’ve already made some changes to the v1alpha5 Provisioner API to support scenarios that could’ve been addressed by a custom launch template. This includes [overriding instanceProfiles](https://github.com/aws/karpenter/pull/914), flexibility to [change maxPods](https://github.com/aws/karpenter/pull/1032), configure [IMDS settings](https://github.com/aws/karpenter/commit/5bb3c3ab4ec840de15f05090761bc5f0733bda46), specify [clusterDNS](https://github.com/aws/karpenter/pull/1013) and more.
+
+The rationale we’ve stuck to so far is
+
+1. We only support basic use cases that should apply to most users. Categorizing a use case as basic has been somewhat arbitrary, but we’ve used upvotes on open issues and PRs to get that signal.
+2. At the minimum, functionality that other AWS provisioners like ManagedNodegroups give you without having to use a launch template should be available through Karpenter as well. Most of the parameters we’ve recently added fit this tenet.
+
+### How do we evolve LT usage?
+
+We’ve considered a few different approaches for this so far.
+
+**Approach 1 - Continue as is. Inline fields as necessary and enforce complete launch templates otherwise.**
```

Comment:
There is a point worth mentioning here is the relationship between kubelet settings, assumptions made by karpenter, and launch templates.
With karpenter-managed LTs, karpenter dictates kubelet settings through LTs matching its own logic. But with provided LTs, there is often a mismatch between karpenter assumptions and actuals. MaxPods is one example, ephemeral disk size, pod overhead, cgroup limits are others.

So for approach 1, I think it is quite important that karpenter is also configurable so that they match relevant kubelet settings and LT parameters.

Reply:
Right, good call let me add that in.  One of the first fields we do intend to make configurable inline is the UserData so the kubelet can always be fully customized without needing a provided LT. 

This is going to be tricky to implement but I expect Karpenter to ignore any setting that would compete with an inline UserData override like specProvider.kubeletConfiguration or even maxPods